zkevm-circuits/src/sig_circuit.rs

```diff
@@ -446,6 +513,202 @@ impl<F: Field> SigCircuit<F> {
+            (sig_is_valid, gate.load_zero(ctx)) // cache zero ?
+        };
+
+        Ok(AssignedECDSA {
+            pk: pk_assigned,
+            pk_is_zero,
+            msg_hash,
+            integer_r,
+            integer_s,
+            v: assigned_y_is_odd,
+            sig_is_valid,
+        })
+    }
+
+    // check if secp256k1 sig.
+    fn is_k1_sig<
+        Fp: PrimeField<Repr = [u8; 32]> + halo2_base::utils::ScalarField,
+        Affine: CurveAffine<Base = Fp>,
+    >(
+        &self,
+    ) -> bool {
+        Affine::a() == Fp::ZERO
+    }
+
+    // check v and y oddness relation for secp2256k1
+    fn check_y_oddness<Fp: PrimeField<Repr = [u8; 32]> + halo2_base::utils::ScalarField>(
+        &self,
+        ctx: &mut Context<F>,
+        ecdsa_chip: &FpConfig<F, Fp>,
+        v: &u8,
+        y_coord: CRTInteger<F>,
+        pk_is_zero: AssignedValue<F>,
+    ) -> (AssignedValue<F>, AssignedValue<F>) {
+        // we constrain:
+        // - v + 2*tmp = y where y is already range checked (88 bits)
+        // - v is a binary
+        // - tmp is also < 88 bits (this is crucial otherwise tmp may wrap around and break
+        //   soundness)
+        let gate = ecdsa_chip.gate();
+        let zero = gate.load_zero(ctx);
+
+        let assigned_y_is_odd = gate.load_witness(ctx, Value::known(F::from(*v as u64)));
+        gate.assert_bit(ctx, assigned_y_is_odd);
+
+        // the last 88 bits of y
+        let assigned_y_limb = &y_coord.limbs()[0];
+        let mut y_value = F::zero();
+        assigned_y_limb.value().map(|&x| y_value = x);
+
+        // y_tmp = (y_value - y_last_bit)/2
+        let y_tmp = (y_value - F::from(*v as u64)) * F::TWO_INV;
+        let assigned_y_tmp = gate.load_witness(ctx, Value::known(y_tmp));
+
+        // y_tmp_double = (y_value - y_last_bit)
+        let y_tmp_double = gate.mul(
+            ctx,
+            QuantumCell::Existing(assigned_y_tmp),
+            QuantumCell::Constant(F::from(2)),
+        );
+        let y_rec = gate.add(
+            ctx,
+            QuantumCell::Existing(y_tmp_double),
+            QuantumCell::Existing(assigned_y_is_odd),
+        );
+        let y_is_ok = gate.is_equal(
+            ctx,
+            QuantumCell::Existing(*assigned_y_limb),
+            QuantumCell::Existing(y_rec),
+        );
+
+        // last step we want to constrain assigned_y_tmp is 87 bits
+        let assigned_y_tmp = gate.select(
+            ctx,
+            QuantumCell::Existing(zero),
+            QuantumCell::Existing(assigned_y_tmp),
+            QuantumCell::Existing(pk_is_zero),
+        );
+
+        println!("assigned_y_tmp {:?}", assigned_y_tmp);
+        //let ecc_chip = EccChip::<F, FpChipK1<F>>::construct(ecdsa_chip.clone());
+
+        let ecc_chip = EccChip::construct(ecdsa_chip.clone());
+        ecc_chip
+            .field_chip
+            .range
+            .range_check(ctx, &assigned_y_tmp, 87);
```

Comment:
a lot of 87/88 in this file.  Make a constant?

Reply:
good point , will update